### PR TITLE
White-listing parameters

### PR DESCRIPTION
**Pull request by Dynom**
Only allow a predefined set of parameters in the URL, e.g.:

```
./proxima -allowed-params=width,height,url
```

With:
```
?width=100&height=100&file=/path/to/image.jpg
```

would result in:
```
error="parameter is not white-listed" parameter=url allowed=width,height,url
```

### CODE REVIEW

File: handlers/allowedparams.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			requestParams := r.URL.Query()
+
+			for p := range requestParams {
+				if _, exists := params[p]; !exists {
+					l.Log("error", "parameter is not white-listed", "parameter", p, "allowed", strings.Join(allowedParams, ","))
+					http.Error(w, "Unregisterd parameter", http.StatusNotAcceptable)
```
Comment (yazgazan) [inline handlers/allowedparams.go]: See comment on #2 

(https://github.com/Dynom/proxima/pull/2/files/b7ea268c1adb92eb875207a99bcc9939ac1661b9#r113132120)

File: main.go [hunk 2]
```diff
@@ -19,11 +19,12 @@ import (
 )
 
 var (
-	allowedHosts argumentList
-	imaginaryURL string
-	listenPort   int64
-	bucketRate   float64
-	bucketSize   int64
+	allowedHosts           argumentList
+	allowedImaginaryParams string
```
Comment (yazgazan) [inline main.go]: See comment on #2 

(https://github.com/Dynom/proxima/pull/2/files/b7ea268c1adb92eb875207a99bcc9939ac1661b9#r113132557)

**Review by yazgazan — CHANGES_REQUESTED** [review]

File: main.go [hunk 3]
```diff
@@ -34,24 +32,13 @@ var (
...
-}
-
-func (l *argumentList) Set(value string) error {
-	*l = append(*l, value)
-	return nil
-}
-
 func init() {
-	flag.Var(&allowedHosts, "allow-host", "Repeatable flag for hosts to allow for the URL parameter (e.g. \"d2dktr6aauwgqs.cloudfront.net\")")
+	flag.Var(&allowedHosts, "allow-host", "Repeatable flag (or a comma-separated list) for hosts to allow for the URL parameter (e.g. \"d2dktr6aauwgqs.cloudfront.net\")")
```
Comment (yazgazan) [inline main.go]: Should be renamed `allow-hosts` (with an ending `s`)

File: argumentlist.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+type argumentList []string
+
+func (l argumentList) String() string {
+	return strings.Join(l, ",")
+}
+
+func (a *argumentList) Set(value string) error {
+	*a = append(*a, strings.Split(value, ",")...)
+	return nil
```
Comment (yazgazan) [inline argumentlist.go:13]: Would returning an error when `value` is empty make sense ? (an empty string would be added to the `argumentList`)

**Review by yazgazan — CHANGES_REQUESTED** [review]

File: argumentlist.go (hunk 4 above)
Comment (Dynom) [reply argumentlist.go:13]: It's not a problem for the implementation, empty values are just ignored during processing. So I wouldn't consider it an error.

**Review by yazgazan — APPROVED** [review]

**Pull request merged**